Min # Jsonparser step3

undervineg opened the pull request:
ㅠㅠ 다 고치고 싶지만.. PR 확인 부탁드립니다.

undervine committed:
Remove trimExceptDataInside() and removeSquareBracket func

Remove funtions that trim raw data except data inside.
Modify JSONObjectPattern and JSONArrayPattern regular expression to find data include whitespaces.

undervine committed:
Add test cases that check data with whitespaces

undervine committed:
Add GrammarChecker struct

Move JsonError from InputView to GrammarChecker.
Add various regular expression statements.

undervine committed:
Add new types of JSONData

Add DataType and a variable dataType that indicates the data is array or object.
Add arrayString, arrayNumber, arrayBool, arrayInner, and objectArray to store array data.
Add initializer for array data.

undervine committed:
Separate regular pattern and add comments

undervine committed:
Add funcs to parse nested data(1depth)

Add makeRealData() to branch process of converting values by the type of data.
Add convertStringToAnyArray() to convert inner arrays or inner objects.
Modify parse() and generateJSONData to return JSONData, not [JSONData].
Add isObjectType() to check user data's type.

undervine committed:
Merge overall report and specific report

Change way to print report. Print the number of arrays and objects treated as basic data.

undervine committed:
Add test cases

Add nested test cases.
Add invalid test cases including whitespaces of object's key, values without commas, values without outermost brackets, etc.

godrm commented on JSONParser/JSONParser/JSONParser.swift:
```diff
@@ -9,114 +9,190 @@
+        // 밸류 생성.
+        let values = try generateValue(from: data, type: .object)
+        // 각 키값과 밸류값을 딕셔너리에 붙임.
+        for i in 0..<keys.count {
```
generateKey와 generateValue 가 결국에는 합쳐서 한꺼번에 dictionary를 만드는게 더 좋지 않을까요? 
두 개로 나눠서 처리한 이유가 있나요?

undervineg replied:
generateValue()는 딕셔너리 외에 배열 데이터를 생성할 때도 따로 쓰이고 있어서 붙이지 않았습니다. 지금 보니 배열 값을 떼어내는 함수를 따로 만드는 게 낫겠네요.

godrm commented on JSONParser/JSONParser/main.swift:
```diff
@@ -12,9 +12,11 @@ func main() throws {
+            // JSON 규격 여부 검사.
+            guard try GrammarChecker.isJSONPattern(userInput) else { throw GrammarChecker.JsonError.invalidPattern }
+            let jsonData = try JSONParser.parse(userInput)
+            try OutputView(jsonData).printReport()
```
OutputView 객체를 활용하는 방법을 OutputView().printReport(jsonData) 형태가 되는게 더 좋을 것 같습니다.
왜냐하면 OutputView 객체 인스턴스를 매번 새로 만들 필요도 없고, 
그러면 내부 변수를 보관할 필요가 없으니 구조도 단순해 질 것 같습니다.

undervineg replied:
네 알겠습니다~!

undervine committed:
Reduce data variables and add setCounts()

Reduce variables that store values of each data types.
Add setCounts() that counts the number of each data.

undervine committed:
Modify some functions

Modify generateKey(), generateValue() to return just a value, to combine each key and value in generateDictionary().

undervine committed:
Add error type 'dataOfNil' in JsonError enum

undervine committed:
Modify printReport() func to static

Modify printReport() to static func.
Modify print process as JSONData struct changes.

undervine committed:
Test some cases

undervine committed:
Add and Remove some comments

undervineg commented:
- generateKey(), generateValue()를 합쳐서 딕셔너리를 만들도록 수정하였습니다.
- OutputView의 printReport()를 static으로 만들었습니다.
- JSONData 내부 변수를 줄였습니다.

undervine committed:
Modify files according to codacy

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -0,0 +1,73 @@
+    // 중첩 배열형 패턴.
+    static let nestedArrayPattern = "\\[" + GrammarChecker.arrayValues + "+\\]"
+    // 배열 내 삽입가능 값 패턴.
+    static let arrayValuePattern = GrammarChecker.basicValuePattern+"|"+GrammarChecker.arrayValues+"|"+GrammarChecker.objectValues
```
private static let 과 static let 구분은 어떤 기준일까요?

undervineg replied:
아 제가 빼먹었네요 ㅎㅎ 수정하겠습니다.

godrm commented on JSONParser/JSONParser/JSONData.swift:
```diff
@@ -9,34 +9,82 @@
 
 struct JSONData {
+    typealias Object = [String:Any]
+    typealias Array = [Any]
```
Array는 이미 스위프트에 있는 타입이니까 이런 경우에 혼동을 줄수도 있습니다. 
기존에 있는 타입과 동일한 이름으로 typealias를 하지는 마세요.

undervineg replied:
네 알겠습니다.

undervine committed:
Add 'private' keyword

undervine committed:
Modify the name of typealias variables

Modify 'Array' to 'JSONArray', 'Object' to 'JSONObject', because 'Array' is existing type in swift.

godrm commented on JSONParser/JSONParserTests/JSONParserTests.swift:
```diff
@@ -11,71 +11,83 @@ import XCTest
+        let parsedData = try JSONParser.parse(data)
+        XCTAssertEqual(parsedData.arrayCount.string, 1)
+        XCTAssertEqual(parsedData.arrayCount.number, 1)
+        XCTAssertEqual(parsedData.arrayCount.bool, 1)
```
이렇게 Assert가 3개 조건이면 string 개수는 맞고 number 개수가 틀려도 이 함수는 실패할꺼고 반대 경우도 실패할 겁니다. 다시 말하면 3가지 조건중에 어느 조건이 실패했는지 찾기 어렵겠죠?
그래서 테스트 함수에는 Assert가 하나만 있는게 좋습니다. 하나의 경우만 테스트하도록 분리해보세요. 테스트함수 이름도 각 경우를 구체적으로 적어주세요.

undervineg replied:
아 그렇군요 알겠습니다. :)

undervine committed:
Separate test functions contains more than one XCTAssert-funcs

godrm commented on JSONParser/JSONParser/GrammarChecker.swift:
```diff
@@ -0,0 +1,73 @@
+    // 종합배열형 패턴.
+    private static let jsonArrayPattern = "(^\\[" + GrammarChecker.jsonObjectPattern + "\\])"
+    // 전체 json 규격.
+    static let jsonPattern = GrammarChecker.nestedObjectPattern + "|" + GrammarChecker.jsonArrayPattern
```
여전히 `static let` 상수가 있네요. 
이렇게 패턴이 경우의 수가 많은데, enum 타입으로 선언하면 어떤 장단점이 있을까요?

undervineg replied:
- 남겨놓은 static let 상수들은 JSONParser 구조체에서 사용하기 때문에 오픈해놓았습니다.
- enum을 쓰게 되면 패턴 타입들을 한 데 모아둘 수 있어 다른 파일에서 사용할 때도 각각의 변수명을 알지 않아도 되는 등 접근이 편리해지는 장점이 있습니다.
- 반면, 이 경우 enum의 rawValue를 String 타입으로 줘야 하는데, enum 내부에서는 rawValue를 리터럴로 써야 하기 때문에 복잡한 정규표현식을 여러 변수로 쪼개서 쓰지 못하고 문자열 자체로만 작성해야 합니다. 그래서 정규표현식을 보고 어떤 의미인지 쉽게 알 지 못하는 단점이 있습니다.

undervineg replied:
그런데 swift에서는 정규표현식의 재귀표현 (?1) 이 작동 안 하는 게 맞나요~?

godrm replied:
정규표현식 재귀표현은 실제로 지원하는 라이브러리에 따라 다릅니다. ICU라고 하는 유명한(?) 오픈소스가 있는데, macOS와 iOS에서도 이걸 사용해서 구현해서 지원하지 않는걸로 알고 있어요.

godrm commented:
다음 단계로 넘어가서 좀 더 고민해봅시다. 수고했습니다
